File: backend/src/main/kotlin/com/alpha/domain/repository/PasswordResetRepository.kt
package com.alpha.domain.repository

import com.alpha.domain.entity.PasswordResetEntity
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import java.time.OffsetDateTime
import java.util.*

interface PasswordResetRepository : JpaRepository<PasswordResetEntity, UUID> {
    fun findByToken(token: String): PasswordResetEntity?
    @Query("SELECT pr FROM PasswordResetEntity pr WHERE pr.user.id = :userId")
    fun findByUserId(@Param("userId") userId: UUID): List<PasswordResetEntity>
    
    @Query("SELECT pr FROM PasswordResetEntity pr WHERE pr.expiresAt < :now")
    fun findExpiredResets(@Param("now") now: OffsetDateTime): List<PasswordResetEntity>
    
    @Modifying
    @Query("DELETE FROM PasswordResetEntity pr WHERE pr.expiresAt < :now")
    fun deleteExpiredResets(@Param("now") now: OffsetDateTime): Int
    
    @Modifying
    @Query("DELETE FROM PasswordResetEntity pr WHERE pr.user.id = :userId")
    fun deleteByUserId(@Param("userId") userId: UUID): Int
}

File: backend-scala-zio/src/main/scala/com/alpha/repository/PasswordResetRepository.scala
package com.alpha.repository

import zio.*
import com.alpha.domain.model.*
import java.util.UUID
import java.time.OffsetDateTime
import io.getquill.*
import io.getquill.extras.OffsetDateTimeOps

trait PasswordResetRepository:
  def findById(id: UUID): Task[Option[PasswordReset]]
  def findByUserId(userId: UUID): Task[List[PasswordReset]]
  def findByToken(token: String): Task[Option[PasswordReset]]
  def findActiveByUserId(userId: UUID): Task[Option[PasswordReset]]
  def create(reset: PasswordReset): Task[UUID]
  def markAsUsed(id: UUID): Task[Int]
  def deleteExpired: Task[Int]

object PasswordResetRepository:
  val layer: ZLayer[PostgresCtx, Nothing, PasswordResetRepository] =
    ZLayer.fromFunction(new PasswordResetRepositoryImpl(_))

class PasswordResetRepositoryImpl(ctx: PostgresCtx) extends PasswordResetRepository:

  import ctx.*

  override def findById(id: UUID): Task[Option[PasswordReset]] =
    ZIO.attempt:
      run(query[PasswordReset].filter(_.id == lift(id))).headOption

  override def findByUserId(userId: UUID): Task[List[PasswordReset]] =
    ZIO.attempt:
      run(query[PasswordReset].filter(_.userId == lift(userId))).toList

  override def findByToken(token: String): Task[Option[PasswordReset]] =
    ZIO.attempt:
      run(query[PasswordReset].filter(_.token == lift(token))).headOption

  override def findActiveByUserId(userId: UUID): Task[Option[PasswordReset]] =
    ZIO.attempt:
      run(query[PasswordReset]
        .filter(r =>
          r.userId == lift(userId) && r.isUsed == lift(false) && r.expiresAt > lift(
            OffsetDateTime.now()))).headOption

  override def create(reset: PasswordReset): Task[UUID] =
    ZIO.attempt:
      run(query[PasswordReset].insertValue(lift(reset)))
      reset.id

  override def markAsUsed(id: UUID): Task[Int] =
    ZIO.attempt:
      run(query[PasswordReset]
        .filter(_.id == lift(id))
        .update(_.isUsed -> lift(true)))
      1

  override def deleteExpired: Task[Int] =
    ZIO.attempt:
      run(query[PasswordReset]
        .filter(_.expiresAt < lift(OffsetDateTime.now()))
        .delete)
      0
